storage/flux/window.go

```diff
@@ -0,0 +1,183 @@
+package storageflux
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/apache/arrow/go/arrow/array"
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/arrow"
+	"github.com/influxdata/flux/execute"
+	"github.com/influxdata/flux/values"
+	"github.com/influxdata/influxdb/v2"
+)
+
+// splitWindows will split a windowTable by creating a new table from each
+// row and modifying the group key to use the start and stop values from
+// that row.
+func splitWindows(ctx context.Context, in flux.Table, f func(t flux.Table) error) error {
+	wts := &windowTableSplitter{
+		ctx: ctx,
+		in:  in,
+	}
+	return wts.Do(f)
+}
+
+type windowTableSplitter struct {
+	ctx context.Context
+	in  flux.Table
+}
+
+func (w *windowTableSplitter) Do(f func(flux.Table) error) error {
+	defer w.in.Done()
+
+	startIdx, err := w.getTimeColumnIndex(execute.DefaultStartColLabel)
+	if err != nil {
+		return err
+	}
+
+	stopIdx, err := w.getTimeColumnIndex(execute.DefaultStopColLabel)
+	if err != nil {
+		return err
+	}
+
+	return w.in.Do(func(cr flux.ColReader) error {
+		// Retrieve the start and stop columns for splitting
+		// the windows.
+		start := cr.Times(startIdx)
+		stop := cr.Times(stopIdx)
+
+		// Iterate through each time to produce a table
+		// using the start and stop values.
+		arrs := make([]array.Interface, len(cr.Cols()))
+		for j := range cr.Cols() {
+			arrs[j] = getColumnValues(cr, j)
+		}
+
+		for i, n := 0, cr.Len(); i < n; i++ {
+			startT, stopT := start.Value(i), stop.Value(i)
+
+			// Rewrite the group key using the new time.
+			key := groupKeyForWindow(cr.Key(), startT, stopT)
+
+			// Produce a slice for each column into a new
+			// table buffer.
+			buffer := arrow.TableBuffer{
+				GroupKey: key,
+				Columns:  cr.Cols(),
+				Values:   make([]array.Interface, len(cr.Cols())),
+			}
+			for j, arr := range arrs {
+				buffer.Values[j] = arrow.Slice(arr, int64(i), int64(i+1))
+			}
+
+			// Wrap these into a single table and execute.
+			done := make(chan struct{})
+			table := &windowTableRow{
+				buffer: buffer,
+				done:   done,
+			}
+			if err := f(table); err != nil {
+				return err
+			}
+
+			select {
+			case <-done:
```

Comment:
Can you explain why a `done` channel is needed (I've seen this in most of the table structures)? I thought the call of `f(table)` will be blocking and synchronous?

Reply:
This is to prevent oversaturation of tables. The `Process` call isn't blocking and synchronous. It gets redirected through the consecutive transport. This prevents the source from continuously producing tables and consuming additional memory and keeps the process synchronous.

I think this detail is confusing because it isn't evident as part of this section of code, but as part of the executor and dispatcher.

Reply:
Oh right, I forgot about the dispatcher.